### ISSUE TITLE
Fine-tuning model does'nt learn while training

### ISSUE THREAD

**Issue by mukeshnarendran7** [yangsenius/TransPose#32]
.  I am using these parameters found in the main git but Fine-tuning model doesn't seem to learn from new data

```
criterion = torch.nn.MSELoss(reduction="mean")
pretrain_part = [param for name, param in model.named_parameters()if 'final_layer' not in name]
optimizer = torch.optim.Adam([ {'params': pretrain_part, 'lr':1e-5 },

```
model loading
```
#Load pre-trained model and fine-tune
model = torch.hub.load('yangsenius/TransPose:main', 
                       'tpr_a4_256x192',
                       pretrained=True)
for param in model.parameters():
    param.requires_grad = False

model.final_layer =  torch.nn.Sequential(torch.nn.Conv2d(256, 18,1))                                    
model = model.to(device)

#Get model summary
summary(model,
          input_size=(3, 256, 192),
          batch_size=1
          )
```
Loss:
Epoch:0, time22.861s, loss0.49620020389556885
Epoch:1, time23.730s, loss0.28467278741300106
Epoch:2, time23.186s, loss0.19562865188345313
Epoch:3, time23.089s, loss0.15849934425204992
Epoch:4, time23.006s, loss0.1431811167858541
Epoch:5, time22.985s, loss0.1366584706120193
Epoch:6, time23.165s, loss0.13364548794925213
Epoch:7, time23.167s, loss0.13212952483445406
Epoch:8, time22.938s, loss0.13107420597225428
Epoch:9, time22.993s, loss0.13025714457035065
Epoch:10, time22.988s, loss0.12950784899294376
Epoch:11, time22.960s, loss0.12873529829084873
Epoch:12, time22.971s, loss0.12812337931245565
Epoch:13, time23.206s, loss0.12742456002160907
Epoch:14, time23.180s, loss0.12706445762887597

**Comment by yangsenius** [comment]
```
for param in model.parameters():
    param.requires_grad = False
```
Why setting requires_grad= False? This will freeze all parameters

**Comment by yangsenius** [comment]
1. Nope. I made very simple modification. Just replace the final layer `1x1 Conv(d, 17)` with `1x1 Conv(d, 16)`. No freezing for previous layers but 1e-5 fixed learning rate for the pretrained layers.
2.  100 epochs
3. Just like this https://github.com/yangsenius/TransPose/issues/31#issuecomment-1041048153. I made very simple changes
4.  Make sure that you have correctly loaded the COCO pretrained weights for the pretrained parts (excluding the final layer) AND correct learning rate settings

**Comment by mukeshnarendran7** [comment]
Hey Thanks for getting back

- >>Load pretrained weights from url: https://github.com/yangsenius/TransPose/releases/download/Hub/tp_r_256x192_enc4_d256_h1024_mh8.pth

100%
22.9M/22.9M [00:00<00:00, 84.4MB/s]

Successfully loaded model  (on cpu) with pretrained weights
- In point 4,  I believe that if i download from torch hub then it should load with the pre-trained coco weights automatically or something else needs to be done

**Comment by mukeshnarendran7** [comment]
Just to share with you on how the model learns when i train on my data the outcomes looks like. Any idea why it might be?
![image](https://user-images.githubusercontent.com/23098703/157662806-e7104f6a-1117-41b1-a470-62e9beed1eba.png)

**Comment by yangsenius** [comment]
Are the positions of the GT keypoints right? From the target heatmap, it seems they are not matched with the input image.

**Comment by mukeshnarendran7** [comment]
Hey, 
The above images are not related but you can see in this one I have plotted the kps over the image and they match and its in all the cases of images that they match
![image](https://user-images.githubusercontent.com/23098703/158051947-1db9a907-254c-4887-b849-37d72157b763.png)

I tried with another dataset and you can see after training for 100 epochs it's not able to generalize to my data.
![image](https://user-images.githubusercontent.com/23098703/158052874-b3dca854-9b7f-44b6-ae45-950c67085956.png)
![image](https://user-images.githubusercontent.com/23098703/158052934-852726c6-5357-4637-86b4-ec85d70e05e8.png)

I tried to use the pre-trained weight to load the model directly to plot on an image from the internet and you can see that it is not able to predict.

```
#Load pre-trained model and fine-tune
model_tp = torch.hub.load('yangsenius/TransPose:main', 
                       'tpr_a4_256x192',
                       pretrained=True)
# for name,param in model_tp.named_parameters():
#     param.requires_grad = True

model = model_tp

#pre-trained mean/std
mean=[0.485, 0.456, 0.406]
std=[0.229, 0.224, 0.225]

trfm = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize(mean=mean,
                            std=std)])
model.eval()
with torch.no_grad():
    
    image = cv2.imread("/content/RN.jpg",1) #image resize
    image = cv2.resize(image, (256, 192))
    temp  = image #for plotting purposes
    image = trfm(image) # normalize, to tensor
    image = image[None].float()
    output = model(image)
    #Get keypoints (used function from yangseius github repo)
    preds, maxvals = get_max_preds(output.clone().cpu().numpy())

    # im = image.squeeze(0).permute(1,2,0), (64, 48)
    x = preds[0][:,0] *(256/64),
    y = preds[0][:,1] *(192/48)
    plt.figure(figsize=(15, 8))
    plt.imshow(temp)
    plt.scatter(x, y, c="r")
    plt.show()
```
Predictions:
![image](https://user-images.githubusercontent.com/23098703/158052736-4a900812-1c34-4080-87f4-73481a4f7595.png)
![image](https://user-images.githubusercontent.com/23098703/158052762-88a032fe-c3de-4c4e-8c82-b8e58d48e9f2.png)

Am I loading the model wrong? I referenced the Pytorch docs and they seem okay. If you write a demo code on how to load it from torch.hub() for pretraining would be helpful.
 
Here's mine
```
#Load pre-trained model and fine-tune
model_tp = torch.hub.load('yangsenius/TransPose:main', 
                       'tpr_a4_256x192',
                       pretrained=True)
for name,param in model_tp.named_parameters():
    param.requires_grad = True
        
model_tp.final_layer =  torch.nn.Sequential(torch.nn.Conv2d(256, 16, kernel_size=1))                                    
model = model_tp.to(device)

#Get model summary
summary(model,
          input_size=(3, 256, 192),
          batch_size=1
          )
```

**Comment by mukeshnarendran7** [comment]
I also modified the heatmaps generator code to something like this by referring to your code. I am not using the visibility of points in my code
```
def adjust_targets(heatmap_size, j_x, j_y, tmp_size):
    mu_x = j_x
    mu_y = j_y
    # Check that any part of the gaussian is in-bounds
    ul = [int(mu_x - tmp_size), int(mu_y - tmp_size)]
    br = [int(mu_x + tmp_size + 1), int(mu_y + tmp_size + 1)]
    if ul[0] >= heatmap_size[0] or ul[1] >= heatmap_size[1] \
            or br[0] < 0 or br[1] < 0:
        # If not, just return the image as is
        mu_x = 0
        mu_y = 0

    return mu_x, mu_y

def get_heatmaps_likelihood(sample, sigma=2,to_size=(64, 48), normalize=True):
    """
        Generates heatmaps from the keypoints of a sample. The std-dev depends on the likelihood of each keypoint. If the likelikhood is high, the std-dev will be smaller.
        :param sample: The sample from which to generate heatmaps.
        :param sigma: The standard deviation of the gaussian noise
        :param normalize: Whether to normalize the heatmaps
        :return: The heatmaps of the keypoints
        """
    image, keypoints = sample['image'], sample['keypoints']
    likelihood = np.ones(np.shape(keypoints)[0])
    kps_xy =[]
    tmp_size = sigma*3 
    for i, (x, y) in enumerate(keypoints):
        #rescale
        # j_x  = x*(48/192)
        # j_y  = y*(64/256)
        x, y = adjust_targets(to_size,
                              j_x, j_y,
                              tmp_size)
        x, y = j_x, j_y
        if x and y ==0:
            continue
            #points may not be annotated or missing
            # x = -1
            # y = -1


        if x>192.0 and y > 256.0:
            raise Exception("keypoints are out of image/bounds")
        if np.isnan(x) or np.isnan(y):
            raise Exception("Nan values present")
            break  
        kps_xy.append([x,y])

    w, h = to_size

    x = np.arange(w)
    y = np.arange(h)
    xx, yy = np.meshgrid(x, y)

    keypoints = kps_xy
    heatmaps = np.zeros([np.shape(keypoints)[0],h, w])
    for i, keypoint in enumerate(keypoints):
        if keypoint[0] < 0 or keypoint[1] < 0:
            continue
        kx = keypoint[0]
        ky = keypoint[1]

        heatmaps[i] = np.exp(-((yy - ky) ** 2 + (xx - kx) ** 2) / (2 * sigma ** 2))

        if not normalize:
            heatmaps[i] /= sigma * np.sqrt(2 * np.pi)

    return heatmaps
```

**Comment by mukeshnarendran7** [comment]
Hey sorry to spam you so many time but i have a good update.


```
#Load pre-trained model and fine-tune
model_tp = torch.hub.load('yangsenius/TransPose:main',
                          'tph_a4_256x192',
                          pretrained=True)
# for name,param in model_tp.named_parameters():
#     param.requires_grad = True

model = model_tp

#pre-trained mean/std
mean=[0.485, 0.456, 0.406]
std=[0.229, 0.224, 0.225]

trfm = transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize(mean=mean,
                            std=std)])
model.eval()
with torch.no_grad():
    
    image = cv2.imread("/content/messi.jpg",1) #image resize
    image = cv2.resize(image, (256, 192))
    temp  = image #for plotting purposes
    image = trfm(image) # normalize, to tensor
    image = image[None].float()
    output = model(image.to('cpu'))
    #Get keypoints (used function from yangseius github repo)
    preds, maxvals = get_max_preds(output.clone().cpu().numpy())

    # im = image.squeeze(0).permute(1,2,0), (64, 48)
    x = preds[0][:,0] *(256/64),
    y = preds[0][:,1] *(192/48)
    plt.figure(figsize=(15, 8))
    plt.imshow(temp)
    plt.scatter(x, y, c="r")
    plt.show()
```
predictions from the other model you have uploaded
![image](https://user-images.githubusercontent.com/23098703/158073517-173b1849-8ca9-4ec7-a278-88b7298deb3c.png)


from my dataset. trained for 30 epochs only and you  can see it doing well also to new images
![image](https://user-images.githubusercontent.com/23098703/158073683-a5bf541f-07db-40d8-896a-6e5e1584efb6.png)

I am starting to think the weights from the a3 might be off. Can you check and see from your side. Thanks

**Comment by yangsenius** [comment]
It seems that your input images are (H=192, W=256) sizes but the pretrained models are trained with (H=256, W=192) sizes. It may be better to make the height of the image large than the width.

**Comment by mukeshnarendran7** [comment]
Hey I tried that initially but the model throws back an error if i load (256, 192) input images
